File: packages/python/src/amplifier_agent/_records.py
"""Public Python values for agent operations and observations."""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from pathlib import Path
from typing import Any, Literal


class AgentError(Exception):
    def __init__(
        self,
        code: str,
        category: str,
        message: str,
        remedy: str,
        retryable: bool = False,
        correlation_id: str | None = None,
        details: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(message)
        self.code = code
        self.category = category
        self.message = message
        self.remedy = remedy
        self.retryable = retryable
        self.correlation_id = correlation_id
        self.details = details

    def __reduce__(self) -> tuple[Any, ...]:
        return (
            type(self),
            (
                self.code,
                self.category,
                self.message,
                self.remedy,
                self.retryable,
                self.correlation_id,
                self.details,
            ),
            vars(self),
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, AgentError):
            return NotImplemented
        return vars(self) == vars(other)


class ToolFailed(Exception):
    """The caller executor authoritatively reports a failed effect."""


class ToolOutcomeUnknown(Exception):
    """The caller executor cannot establish whether its effect completed."""


@dataclass
class TextPart:
    text: str
    type: Literal["text"] = "text"


ContentPart = TextPart


@dataclass
class ConversationMessage:
    role: Literal["system", "developer", "user", "assistant"]
    content: list[ContentPart]


@dataclass
class TurnInput:
    content: list[ContentPart]
    model: str | None = None
    history: list[ConversationMessage] | None = None


@dataclass
class UsageEntry:
    provider: str
    model: str
    tokens_in: int | None = None
    tokens_out: int | None = None
    cache_read_tokens: int | None = None
    cache_write_tokens: int | None = None
    cost: dict[str, Decimal] | None = None


@dataclass
class Usage:
    entries: list[UsageEntry]


@dataclass
class TurnResult:
    state: Literal["success", "failure", "rejected", "cancelled"]
    content: list[ContentPart] | None = None
    error: AgentError | None = None
    usage: Usage | None = None


@dataclass(frozen=True)
class SessionRecord:
    session_id: str
    persistence: Literal["durable", "ephemeral"]


@dataclass(frozen=True)
class TurnInfo:
    session_id: str
    turn_id: str


@dataclass
class TurnRecord:
    turn_id: str
    input: TurnInput
    result: TurnResult


@dataclass
class Event:
    contract_version: str
    session_id: str
    turn_id: str
    sequence: int
    type: str
    payload: Any
    at: datetime | None = None


@dataclass(frozen=True)
class ToolContext:
    call_id: str
    deadline: datetime | None = None


ToolHandler = Callable[[dict[str, Any], ToolContext], Awaitable[str]]


@dataclass
class Tool:
    name: str
    description: str
    input_schema: dict[str, Any]
    handler: ToolHandler
    safety: dict[str, Any] | None = None


@dataclass
class ToolCall:
    call_id: str
    name: str
    source: Literal["built-in", "caller", "mcp"]
    arguments: dict[str, Any]
    deadline: datetime | None = None


@dataclass
class ToolResolution:
    call_id: str
    outcome: Literal["completed", "failed", "cancelled", "unknown"]
    content: str | None = None
    error: AgentError | None = None


@dataclass
class McpServer:
    name: str
    transport: Literal["stdio", "http"]
    command: str | None = None
    args: list[str] | None = None
    env: dict[str, str] | None = None
    url: str | None = None
    headers: dict[str, str] | None = None


@dataclass
class ApprovalRequest:
    request_id: str
    summary: str
    call_id: str | None = None
    name: str | None = None


@dataclass
class ApprovalResponse:
    decision: Literal["allow", "deny", "cancel"]
    reason: str | None = None


ApprovalHandler = Callable[[ApprovalRequest], Awaitable[ApprovalResponse]]


@dataclass
class AgentOptions:
    provider: str | None = None
    model: str | None = None
    instructions: str | None = None
    tools: list[Tool] | None = None
    skills: list[str] | None = None
    mcp_servers: list[McpServer] | None = None
    storage: str | Path | None = None
    approvals: ApprovalHandler | Literal["allow", "deny"] | None = None
    tool_error_policy: Literal["stop", "continue"] = "stop"


@dataclass
class SessionOptions:
    session_id: str | None = None
    persistence: Literal["durable", "ephemeral"] = "durable"
    model: str | None = None


@dataclass
class Selection:
    provider: str
    model: str


@dataclass
class TurnStarted:
    continuation: Literal["fresh", "resumed"]
    primary_actual: Selection


@dataclass
class OutputDelta:
    content: list[ContentPart]


@dataclass
class ReasoningDelta:
    text: str


@dataclass
class ReasoningFinal:
    text: str


@dataclass
class ToolCallEvent:
    call: ToolCall


@dataclass
class ToolResultEvent:
    resolution: ToolResolution


@dataclass
class ApprovalRequestEvent:
    request: ApprovalRequest


@dataclass
class ApprovalResolution:
    request_id: str
    decision: str
    reason: str | None = None


@dataclass
class ApprovalDecision:
    resolution: ApprovalResolution


@dataclass
class Progress:
    data: Any


@dataclass
class UsageEvent:
    snapshot: Usage

File: packages/typescript/src/records.ts
import type { AgentError } from "./errors.js";

export interface TextPart { type: "text"; text: string }
export type ContentPart = TextPart;
export interface ConversationMessage {
  role: "system" | "developer" | "user" | "assistant";
  content: ContentPart[];
}
export interface TurnInput { content: ContentPart[]; model?: string; history?: ConversationMessage[] }
export interface TurnResult {
  state: "success" | "failure" | "rejected" | "cancelled";
  content?: ContentPart[];
  error?: AgentError;
  usage?: Usage;
}
export interface SessionRecord { session_id: string; persistence: "durable" | "ephemeral" }
export interface TurnInfo { session_id: string; turn_id: string }
export interface TurnRecord { turn_id: string; input: TurnInput; result: TurnResult }
export interface UsageEntry {
  provider: string;
  model: string;
  tokens_in?: bigint;
  tokens_out?: bigint;
  cache_read_tokens?: bigint;
  cache_write_tokens?: bigint;
  cost?: Record<string, string>;
}
export interface Usage { entries: UsageEntry[] }

export interface ToolContext { readonly call_id: string; readonly deadline?: string }
export type ToolHandler = (args: Record<string, unknown>, context: ToolContext) => Promise<string>;
export interface Tool {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
  handler: ToolHandler;
  safety?: Record<string, unknown>;
}
export interface ToolCall {
  call_id: string;
  name: string;
  source: "built-in" | "caller" | "mcp";
  arguments: Record<string, unknown>;
  deadline?: string;
}
export interface ToolResolution {
  call_id: string;
  outcome: "completed" | "failed" | "cancelled" | "unknown";
  content?: string;
  error?: AgentError;
  truncated?: boolean;
  original_bytes?: number;
}
export type McpServer =
  | { name: string; transport: "stdio"; command: string; args?: string[]; env?: Record<string, string> }
  | { name: string; transport: "http"; url: string; headers?: Record<string, string> };
export interface ApprovalRequest {
  readonly request_id: string;
  readonly summary: string;
  readonly call_id?: string;
  readonly name?: string;
}
export interface ApprovalResponse { decision: "allow" | "deny" | "cancel"; reason?: string }
export type ApprovalHandler = (request: ApprovalRequest) => Promise<ApprovalResponse>;
export interface ApprovalResolution {
  request_id: string;
  decision: "allow" | "deny" | "cancel" | "timeout" | "unavailable" | "invalid";
  reason?: string;
}
export interface AgentOptions {
  provider?: string;
  model?: string;
  instructions?: string;
  tools?: Tool[];
  skills?: string[];
  mcpServers?: McpServer[];
  storage?: string;
  approvals?: ApprovalHandler | "allow" | "deny";
  toolErrorPolicy?: "stop" | "continue";
  toolResultMaxBytes?: number | null;
}
export interface SessionOptions { sessionId?: string; persistence?: "durable" | "ephemeral"; model?: string }

interface EventEnvelope<T extends string, P> {
  contract_version: "turn-events/1";
  session_id: string;
  turn_id: string;
  sequence: bigint;
  type: T;
  payload: P;
  at?: string;
}
export interface TurnStarted { continuation: "fresh" | "resumed"; primary_actual: { provider: string; model: string } }
export interface OutputDelta { content: ContentPart[] }
export interface ReasoningDelta { text: string }
export interface ReasoningFinal { text: string }
export interface ToolCallEvent { call: ToolCall }
export interface ToolResultEvent { resolution: ToolResolution }
export interface ApprovalRequestEvent { request: ApprovalRequest }
export interface ApprovalDecision { resolution: ApprovalResolution }
export interface Progress { data: unknown }
export interface UsageEvent { snapshot: Usage }
export type Event =
  | EventEnvelope<"turn_started", TurnStarted>
  | EventEnvelope<"output_delta", OutputDelta>
  | EventEnvelope<"reasoning_delta", ReasoningDelta>
  | EventEnvelope<"reasoning_final", ReasoningFinal>
  | EventEnvelope<"tool_call", ToolCallEvent>
  | EventEnvelope<"tool_result", ToolResultEvent>
  | EventEnvelope<"approval_request", ApprovalRequestEvent>
  | EventEnvelope<"approval_decision", ApprovalDecision>
  | EventEnvelope<"progress", Progress>
  | EventEnvelope<"usage", UsageEvent>
  | EventEnvelope<"terminal", TurnResult>
  | EventEnvelope<`${string}.${string}`, unknown>;

export interface Agent extends AsyncDisposable {
  createSession(options?: SessionOptions): Promise<Session>;
  resumeSession(sessionId: string): Promise<Session>;
  listSessions(): Promise<SessionRecord[]>;
  deleteSession(sessionId: string): Promise<void>;
  close(): Promise<void>;
}
export interface Session extends AsyncDisposable {
  readonly info: SessionRecord;
  readonly history: TurnRecord[];
  run(input: TurnInput): Promise<TurnResult>;
  startTurn(input: TurnInput): Promise<Turn>;
  fork(): Promise<Session>;
  close(): Promise<void>;
}
export interface Turn {
  readonly info: TurnInfo;
  events(): AsyncIterable<Event>;
  cancel(): Promise<void>;
}
